src/rust/engine/process_execution/src/docker.rs

```diff
@@ -0,0 +1,330 @@
+    };
+
+    log::debug!("creating container with config: {:?}", &config);
+
+    let container = self
+      .docker
+      .create_container::<&str, String>(None, config)
+      .await
+      .map_err(|err| format!("Failed to create Docker container: {:?}", err))?;
+
+    log::debug!("created container {}", &container.id);
+
+    // DOCKER-TODO: Consider adding a drop guard to remove the container on error? (Although
+    // auto-remove has been disabled if self.keep_sandboxes is "always" or "on failure.")
+
+    self
+      .docker
+      .start_container::<String>(&container.id, None)
+      .await
+      .map_err(|err| {
+        format!(
+          "Failed to start Docker container `{}`: {:?}",
+          &container.id, err
+        )
+      })?;
+
+    log::debug!("started container {}", &container.id);
+
+    let attach_options = bollard::container::AttachContainerOptions::<String> {
+      stdout: Some(true),
+      stderr: Some(true),
+      logs: Some(true), // stream any output that was missed between the start_container call and now
+      stream: Some(true),
+      ..bollard::container::AttachContainerOptions::default()
+    };
+
+    let attach_result = self
+      .docker
+      .attach_container(&container.id, Some(attach_options))
+      .await
+      .map_err(|err| {
+        format!(
+          "Failed to attach to Docker container `{}`: {:?}",
+          &container.id, err
+        )
+      })?;
+
+    log::debug!("attached to container {}", &container.id);
+
+    let mut output_stream = attach_result.output.boxed();
+
+    let wait_options = bollard::container::WaitContainerOptions {
+      condition: "not-running",
+    };
+    let mut wait_stream = self
+      .docker
+      .wait_container(&container.id, Some(wait_options))
+      .boxed();
+
+    let container_id = container.id.to_owned();
+    let result_stream = async_stream::stream! {
+      let mut run_stream = true;
+      while run_stream {
+        tokio::select! {
+          Some(output_msg) = output_stream.next() => {
+            match output_msg {
+              Ok(LogOutput::StdOut { message }) => {
+                log::debug!("container {} wrote {} bytes to stdout", &container_id, message.len());
+                yield Ok(ChildOutput::Stdout(message));
+              }
+              Ok(LogOutput::StdErr { message }) => {
+                log::debug!("container {} wrote {} bytes to stderr", &container_id, message.len());
+                yield Ok(ChildOutput::Stderr(message));
+              }
+              _ => (),
+            }
+          }
+          Some(wait_msg) = wait_stream.next() => {
+            log::debug!("wait_container stream ({}): {:?}", &container_id, wait_msg);
+            match wait_msg {
+              Ok(r) => {
+                // DOCKER-TODO: How does Docker distinguish signal versus exit code? Improve
+                // `ChildResults` to better support exit code vs signal vs error message?
```

Comment:
Usually a negative exit code is a signal: it looks like `status_code` is `i64` for them, so I expect it's similar.